prow/cmd/webhook-server/main.go

```diff
@@ -48,47 +47,125 @@ type ClientInterface interface {
+		logrus.WithError(err).Fatal("Error getting kubeconfig")
 	}
+	var certFile string
+	var privKeyFile string
+	var tempDir string
 	ctx := context.Background()
-	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
+	cl, err := ctrlruntimeclient.New(kubeCfg, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not create writer client")
+	}
+	var client ClientInterface
+	if len(o.projectId) > 0 {
+		client, err = secretmanager.NewClient(o.projectId, false)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create secret manager client")
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	if len(o.fileSystemPath) > 0 {
+		absPath, err := filepath.Abs(o.fileSystemPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to generate absolute file path")
+		}
+		client = &FileSystemClient{
+			path:   absPath,
+			expiry: o.expiryInYears,
+			dns:    o.dnsNames,
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	http.HandleFunc("/validate", serveValidate)
+	logrus.Info("Listening on port 8008...")
+	logrus.Fatal(http.ListenAndServeTLS(":8008", certFile, privKeyFile, nil))
+}
+
+func handleSecrets(client ClientInterface, ctx context.Context, o options, cl ctrlruntimeclient.Client) (string, string, string) {
+	cert, privKey, _, err := getGCPSecrets(client, ctx, o.expiryInYears, o.dnsNames)
+	var caPem string
 	if err != nil && strings.Contains(err.Error(), gcpSecretError) {
-		logrus.Infof("%v, Will now proceed to create GCP Secret", err)
-		cert, privKey, err = createGCPSecret(client, ctx, exp, dns.Strings())
+		logrus.Infof("%v, Will now proceed to create secrets", err)
+		cert, privKey, caPem, err = createGCPSecret(client, ctx, o.expiryInYears, o.dnsNames)
 		if err != nil {
 			logrus.WithError(err).Fatal("Unable to create ca certificate")
 		}
+		err := createValidatingWebhookConfig(ctx, caPem, cl)
+		if err != nil {
+			logrus.WithError(err).Fatalf("Unable to generate ValidationWebhookConfig")
+		}
 	} else if err != nil {
-		logrus.WithError(err).Fatal("Unable to get GCP Secret")
+		logrus.WithError(err).Fatal("Unable to get secrets")
 	}
 	if err = isCertValid(cert); err != nil {
 		logrus.WithError(err).Info("Certificate is not valid, will replace.")
-		cert, privKey, err = updateGCPSecret(client, ctx, exp, dns.Strings())
+		cert, privKey, caPem, err = updateGCPSecret(client, ctx, o.expiryInYears, o.dnsNames)
 		if err != nil {
 			logrus.WithError(err).Fatal("Unable to update GCP Secret")
 		}
-	} 
+		err := patchValidatingWebhookConfig(ctx, caPem, cl)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to generate ValidationWebhookConfig")
+		}
+	}
 	tempDir, err := ioutil.TempDir("", "cert")
 	if err != nil {
 		logrus.WithError(err).Fatal("Unable to create temp directory")
 	}
-	defer os.RemoveAll(tempDir)
 	certFile := filepath.Join(tempDir, "certFile.pem")
```

Comment:
repeatedly referring to filenames such as `certFile.pem` doesn't feel very reliable. A Go way of doing this, could be:

```
type certFilesBundle struct {
  CertFile []byte `json:"certFile.pem"`
  ...
}

var cfb certFilesBundle
cfb.CertFile = <Content of the file>
...

// Marshal
content, err := json.Marshal(cfb)

// Unmarshal
var gotCfb certFilesBundle
json.Unmarshal(contentDownloaded, &gotCfb)
```

Reply:
im only storing the filenames, since i have to read the content from the filepath to ensure no errors, do i need then to write content to the file, marshal and pass it around?